sound/soc/sof/intel/hda-ipc.c

```diff
@@ -274,6 +276,27 @@ void hda_ipc_msg_data(struct snd_sof_dev *sdev,
 	}
 }
 
+void hda_ipc_receive_msg(struct snd_sof_dev *sdev, u32 msg_cmd)
+{
+	struct sof_intel_hda_dev *hdev = sdev->pdata->hw_pdata;
+	struct sof_ipc_comp_event event;
+	u32 cmd;
+
+	cmd = msg_cmd & SOF_GLB_TYPE_MASK;
+
+	switch (cmd) {
+	case SOF_IPC_GLB_COMP_MSG:
+		snd_sof_ipc_msg_data(sdev, NULL, &event, sizeof(event));
+
+		/* reschedule D0i3 work if keyword detected */
+		if (event.event_type == SOF_CTRL_EVENT_KD)
+			mod_delayed_work(system_wq, &hdev->d0i3_work,
```

Comment:
I'm looking at 63e51fd33fef ("ASoC: SOF: Intel: cnl: Implement feature to support DSP D0i3 in S0") and I'm not sure I understand: why do we need an additional delayed work, why cannot we use runtime-PM to switch over to D0i3? Is it because when any stream is active we don't runtime-put? Maybe we should change that to du a `put()` when only kwd is active? Doing that would already close some potential races with the additional context of the new delayed work. I think this is also related to #2663 and as discussed there it also seems to me, that we need firmware cooperation for this too, don't we?

Reply:
@lyakh pm-runtime only deals with D0 and D3 states. D0i3 is a substate of D0, and from a pm-runtime perspective the state is 'active'. So we need to have a dedicated idleness detection, separate from pm-runtime.

Reply:
@plbossart Hm, interesting. This seems to be a part or a larger topic of multiple fine-grained power states. Runtime PM only handles two states: active and idle. But actual hardware can have finer grade states, like in our case. But doing this completely in the driver with custom timers doesn't seem quite right to me, I'd expect tighter integration with the (rt-)pm subsystem. Maybe there has to be an API in rt-pm to support custom power states, that would mostly just pass implementation to driver callbacks but use a common infrastructure to guarantee consistency and protect against races with the global pm? Maybe it's related to the `wakeup_source_register()` API? With only the trigger phrase active there's actually no streaming or no other host<->device IO happening, but once a trigger phrase is detected, the host has to be woken up? IIUC that's rather a *system* wakeup API, i.e. waking up from a system suspend, which isn't our case. But maybe a similar API is needed for rt-pm?

Reply:
@lyakh it would certainly be useful if runtime-pm was extended to deal with 'active-standby' cases, but at the same time the D0ix states are very very custom, and one could argue that you need to much device-specific information before you can enter this mode that a custom detection is acceptable. Take the example of SOF, we only enter this mode when specific streams tagged in the topology are used.

At any rate, it is what it is today, we need a custom timer for 2021 products.